### DISCUSSION TITLE
Add "x-stop" directive to prevent further DOM walking

### DISCUSSION THREAD

**Discussion by skttl** [alpinejs/alpine#1036] [3. Feature ideas]
As stated in https://ryangjchandler.co.uk/articles/5-simple-alpinejs-tips-and-tricks#1-keep-your-components-small, large components, like adding `x-data` to `<body>` can be a performance problem. Would it help if we had something like `x-stop` to stop walking through specific DOM nodes?

A typical usecase I have is this

```html
<body x-data="{ usingKeyboard: false, locked: false }" 
      x-on:mousedown="usingKeyboard = false" 
      x-on:keydown.tab="usingKeyboard = true" 
      x-on:body-lock="locked = $event.detail" 
      class="body ~ min-h-screen flex-col text-gray-900" x-bind:class="{
         'using-keyboard': usingKeyboard,
         'overflow-hidden': locked === true
      }">
```

I use this to determine if the user is using the keyboard for navigating (so I can enhance the focus styles), or if a modal/dialog/whatever is visible, so I don't want the page to scroll beneath.

In this I could then add something like `<div x-stop`, to have Alpine not walk the contents of that div. My complete markup could then be

```html
<body x-data="...">
<div x-stop><!-- this is not considered part of the component in <body> --></div>
</body>
```

**Comment by KevinBatdorf** [comment]
That doesn't need to go on the body. Events bubble up to the window and you can watch them from anywhere. 

Demo: https://codepen.io/KevinBatdorf/pen/BajqgYJ

**Reply by skttl** [reply]
Yeah, but that doesn't really help when I need to conditionally set overflow-hidden on the body.

And I want to have a keyboard-focus: variant in tailwind, so I don't have to turn all focusable elements into Alpine components.

**Reply by KevinBatdorf** [reply]
In that case you should add the class to the body with Vanilla JS. You could have something like this:

```html
<body>
  <!-- Your other elements -->
  <div x-data @keydown.window.tab="document.body.classList.add('using-keyboard')"></div>
</body>
```

If you're building a modal, you can use a watcher to handle the state: https://codepen.io/KevinBatdorf/pen/PoZZeox

**Reply by skttl** [reply]
That is clever - thanks for taking your time!

I'll adjust to this, and get back if I find any other possible usecases for x-data on html or body :)

**Comment by philwolstenholme** [comment]
Just to add, this sort of idea was previously discussed in https://github.com/alpinejs/alpine/pull/395 and https://github.com/alpinejs/alpine/issues/393 but the outcome was that it wouldn't be implemented unless people could demonstrate more of a need.
